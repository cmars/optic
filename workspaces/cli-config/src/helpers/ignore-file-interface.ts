import fs from 'fs-extra';
import path from 'path';
import { readApiConfig } from '../index';
import lockfile from 'proper-lockfile';

interface IIgnoreFileParsed {
  ruleLines: IIgnoreFileRuleLine[];
  rulesFromYamlDeprecated: string[];
  allRules: string[];
}

interface IIgnoreFileRuleLine {
  line: number;
  rule: string;
}

export class IgnoreFileHelper {
  constructor(private ignoreFilePath: string, private opticYamlPath: string) {}

  async getCurrentIgnoreRules(): Promise<IIgnoreFileParsed> {
    const ignoreFile = path.resolve(this.ignoreFilePath);
    await fs.ensureFile(ignoreFile);

    const contents = await safeRead(this.ignoreFilePath);
    const lines = splitLines(contents);

    const rules: IIgnoreFileRuleLine[] = [];
    lines.forEach((line, index) => {
      if (line.trim() !== '' && !line.trim().startsWith('#')) {
        rules.push({
          line: index,
          rule: line,
        });
      }
    });

    const rulesFromYamlDeprecated = await this.loadDeprecatedRules();
    return {
      ruleLines: rules,
      rulesFromYamlDeprecated,
      allRules: [...rules.map((i) => i.rule), ...rulesFromYamlDeprecated],
    };
  }

  async loadDeprecatedRules(): Promise<string[]> {
    try {
      const config = await readApiConfig(this.opticYamlPath);
      const ignores = config.ignoreRequests || [];
      return ignores;
    } catch (e) {
      return [];
    }
  }

  async appendRule(rule: string): Promise<void> {
    const contents = await safeRead(this.ignoreFilePath);
    await safeWrite(this.ignoreFilePath, contents.trim() + '\n' + rule);
  }
}

async function safeRead(filePath: string): Promise<string> {
  await lockfile.lock(filePath, {
    retries: { retries: 10 },
  });

  const contents = (await fs.readFile(filePath)).toString();
  await lockfile.unlock(filePath);
  return contents;
}

async function safeWrite(filePath: string, contents: string): Promise<void> {
  await lockfile.lock(filePath, {
    retries: { retries: 10 },
  });

  await fs.writeFile(filePath, contents);
  await lockfile.unlock(filePath);
}

function splitLines(t: string): string[] {
  return t.split(/\r\n|\r|\n/);
}
