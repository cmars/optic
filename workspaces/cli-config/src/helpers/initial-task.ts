//@ts-ignore
import jsesc from 'jsesc';

export function buildTask(
  apiName: string,
  flags: any,
  taskName: string
): string {
  const config = `
name: ${escapeIt(apiName)}
# Start your api with Optic by running 'api run <taskname>'
tasks:
${buildInitialTask(flags, taskName)}

# Capture traffic from a deployed api by running 'api intercept <environment-name>'
# pass '--chrome' to capture from your browser's network tab
environments:
  production:
    host: https://api.github.com # the hostname of the API we should record traffic from
    webUI: https://api.github.com/repos/opticdev/optic # the url that should open when a browser flag is passed
`.trimLeft();

  return config;
}

export const defaultCommandInit =
  'echo "Setup A Valid Command to Start your API!"';

function buildInitialTask(flags: any, taskName: string) {
  //default config and valid for start injected
  let commandConfig = `  ${taskName}:
     command: ${escapeIt(flags.command || defaultCommandInit)}
     inboundUrl: ${flags.inboundUrl || 'http://localhost:4000'}
`.trimRight();

  if (flags.inboundUrl && flags.targetUrl) {
    commandConfig = `  ${taskName}:
     inboundUrl: ${flags.inboundUrl}
     targetUrl: ${flags.targetUrl}
`.trimRight();
  }

  return commandConfig;
}

function escapeIt(value: string): string {
  const escaped = jsesc(value, { quotes: 'double' });
  if (escaped !== value) {
    return `"${escaped}"`;
  }
  return `"${value}"`;
}
