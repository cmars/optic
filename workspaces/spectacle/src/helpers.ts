import { shapes, endpoints } from '@useoptic/graph-lib';

export function buildEndpointsGraph(spec: any, opticEngine: any) {
  const serializedGraph = JSON.parse(
    opticEngine.get_endpoints_projection(spec)
  );
  const { nodes, edges, nodeIndexToId } = serializedGraph;

  const indexer = new endpoints.GraphIndexer();

  function remapId(arrayIndex: number) {
    const fallbackId = arrayIndex.toString();
    const id = nodeIndexToId[fallbackId];
    if (id !== undefined) {
      return id;
    }
    return fallbackId;
  }

  nodes.forEach((node: endpoints.Node, index: number) => {
    const id = remapId(index);
    indexer.addNode({
      ...node,
      id,
    });
  });
  edges.forEach((e: [number, number, any]) => {
    const [sourceIndex, targetIndex, edge] = e;
    indexer.addEdge(edge, remapId(sourceIndex), remapId(targetIndex));
  });
  const queries = new endpoints.GraphQueries(indexer);
  return queries;
}

export function buildShapesGraph(spec: any, opticEngine: any) {
  const serializedGraph = JSON.parse(opticEngine.get_shapes_projection(spec));
  const { nodes, edges, nodeIndexToId } = serializedGraph;
  // console.log('nodes', nodes);

  const indexer = new shapes.GraphIndexer();

  function remapId(arrayIndex: number) {
    const fallbackId = arrayIndex.toString();
    const id = nodeIndexToId[fallbackId];
    if (id !== undefined) {
      return id;
    }
    return fallbackId;
  }

  nodes.forEach((node: shapes.Node, index: number) => {
    const id = remapId(index);
    indexer.addNode({
      ...node,
      id,
    });
  });
  edges.forEach((e: [number, number, any]) => {
    const [sourceIndex, targetIndex, edge] = e;
    indexer.addEdge(edge, remapId(sourceIndex), remapId(targetIndex));
  });
  const queries = new shapes.GraphQueries(indexer);
  return queries;
}

type EndpointChange = {
  change: {
    category: string;
  };
  pathId: string;
  path: string;
  method: string;
};

type EndpointChanges = {
  data: {
    endpoints: EndpointChange[];
  };
};

export function buildEndpointChanges(
  endpointQueries: endpoints.GraphQueries,
  shapeQueries: shapes.GraphQueries,
  sinceBatchCommitId?: string
): EndpointChanges {
  // Sorted in reverse order - the latest change takes precedence
  const sortedBatchCommits = endpointQueries
    .listNodesByType(endpoints.NodeType.BatchCommit)
    .results.sort((a: any, b: any) => {
      return a.result.data.createdAt < b.result.data.createdAt ? 1 : -1;
    }) as endpoints.BatchCommitNodeWrapper[];

  let deltaBatchCommits: endpoints.BatchCommitNodeWrapper[];

  if (sinceBatchCommitId) {
    const sinceBatchCommit: any = endpointQueries.findNodeById(
      sinceBatchCommitId!
    );
    deltaBatchCommits = sortedBatchCommits.filter(
      (batchCommit: any) =>
        batchCommit.result.data.createdAt >
        sinceBatchCommit!.result.data.createdAt
    );
  } else {
    deltaBatchCommits = sortedBatchCommits;
  }

  const changes = new Changes();

  deltaBatchCommits.forEach((batchCommit: endpoints.BatchCommitNodeWrapper) => {
    // Only createdIn and removedIn edges pointing to a request is treated as
    // an added or removed change, everything else is updated.
    // Bodies, and shape changes are handled below
    batchCommit
      .createdInEdgeNodes()
      .results.forEach((node: endpoints.NodeWrapper) => {
        if (node.result.type === endpoints.NodeType.Request) {
          changes.captureChange(
            'added',
            endpointFromRequest(node as endpoints.RequestNodeWrapper)
          );
        } else if (node.result.type === endpoints.NodeType.Response) {
          changes.captureChange(
            'updated',
            endpointFromResponse(node as endpoints.ResponseNodeWrapper)
          );
        }
      });
    batchCommit
      .removedInEdgeNodes()
      .results.forEach((node: endpoints.NodeWrapper) => {
        if (node.result.type === endpoints.NodeType.Request) {
          changes.captureChange(
            'removed',
            endpointFromRequest(node as endpoints.RequestNodeWrapper)
          );
        } else if (node.result.type === endpoints.NodeType.Response) {
          changes.captureChange(
            'updated',
            endpointFromResponse(node as endpoints.ResponseNodeWrapper)
          );
        }
      });
    batchCommit
      .updatedInEdgeNodes()
      .results.forEach((node: endpoints.NodeWrapper) => {
        if (node.result.type === endpoints.NodeType.Request) {
          changes.captureChange(
            'updated',
            endpointFromRequest(node as endpoints.RequestNodeWrapper)
          );
        } else if (node.result.type === endpoints.NodeType.Response) {
          changes.captureChange(
            'updated',
            endpointFromResponse(node as endpoints.ResponseNodeWrapper)
          );
        }
      });
  });

  // Gather batch commit neighbors
  const batchCommitNeighborIds = new Map();
  deltaBatchCommits.forEach((batchCommit: any) => {
    const batchCommitId = batchCommit.result.id;
    // TODO: create query for neighbors of all types
    shapeQueries
      .listIncomingNeighborsByType(batchCommitId, shapes.NodeType.Shape)
      .results.forEach((shape: any) => {
        batchCommitNeighborIds.set(shape.result.id, batchCommitId);
      });
    shapeQueries
      .listIncomingNeighborsByType(batchCommitId, shapes.NodeType.Field)
      .results.forEach((field: any) => {
        batchCommitNeighborIds.set(field.result.id, batchCommitId);
      });
  });

  endpointQueries
    .listNodesByType(endpoints.NodeType.Body)
    .results.reduce((results: string[], bodyNode: any) => {
      const { rootShapeId } = bodyNode.result.data;
      if (batchCommitNeighborIds.has(rootShapeId)) {
        results.push(rootShapeId);
        return results;
      }
      for (const descendant of shapeQueries.descendantsIterator(rootShapeId)) {
        if (batchCommitNeighborIds.has(descendant.id)) {
          results.push(rootShapeId);
          return results;
        }
      }
      return results;
    }, [])
    .forEach((changedRootShapeId: any) => {
      const body: any = endpointQueries.findNodeById(changedRootShapeId);
      const response = body.response();
      if (response) {
        if (changes.captureChange('updated', endpointFromResponse(response))) {
          return;
        }
      }
      const request = body.request();
      if (request) {
        changes.captureChange('updated', endpointFromRequest(request));
      }
    });

  return changes.toEndpointChanges();
}

type Endpoint = {
  endpointId: string;
  pathId: string;
  path: string;
  method: string;
};

class Changes {
  public changes: Map<string, EndpointChange>;

  constructor() {
    this.changes = new Map();
  }

  captureChange(
    category: 'added' | 'updated' | 'removed',
    endpoint: Endpoint
  ): boolean {
    if (this.changes.has(endpoint.endpointId)) return false;
    this.changes.set(endpoint.endpointId, {
      change: { category },
      pathId: endpoint.pathId,
      path: endpoint.path,
      method: endpoint.method,
    });
    return true;
  }

  toEndpointChanges(): EndpointChanges {
    return {
      data: {
        endpoints: Array.from(this.changes.values()),
      },
    };
  }
}

function endpointFromRequest(request: endpoints.RequestNodeWrapper): Endpoint {
  let pathNode = request.path();
  const pathId = pathNode.value.pathId;
  const path = pathNode.absolutePathPatternWithParameterNames;
  const method = request.value.httpMethod;
  const endpointId = JSON.stringify({ path, method });
  return { endpointId, pathId, path, method };
}

function endpointFromResponse(
  response: endpoints.ResponseNodeWrapper
): Endpoint {
  let pathNode = response.path();
  const pathId = pathNode.value.pathId;
  const path = pathNode.absolutePathPatternWithParameterNames;
  const method = response.value.httpMethod;
  const endpointId = JSON.stringify({ path, method });
  return { endpointId, pathId, path, method };
}

//@TODO remove if not needed after testing
export function getShapeChanges(
  shapeQueries: shapes.GraphQueries,
  shapeId: string,
  sinceBatchCommitId?: string
): ChangeResult {
  const results = {
    added: false,
    changed: false,
  };

  // TODO: figure out why shapeId is undefined
  if (!shapeId) return results;

  const sinceBatchCommit: any = shapeQueries.findNodeById(sinceBatchCommitId!)!;
  const shape: any = shapeQueries.findNodeById(shapeId)!;
  const deltaBatchCommits = getDeltaBatchCommits(
    shapeQueries,
    sinceBatchCommit.result.Id
  );

  for (const batchCommit of shape.batchCommits().results) {
    if (deltaBatchCommits.has(batchCommit.result.id)) {
      return { ...results, added: true };
    }
  }

  return results;
}

export function getFieldChanges(
  shapeQueries: shapes.GraphQueries,
  fieldId: string,
  shapeId: string,
  sinceBatchCommitId?: string
): ChangeResult {
  const results = {
    added: false,
    changed: false,
  };

  const deltaBatchCommits = getDeltaBatchCommits(
    shapeQueries,
    sinceBatchCommitId
  );

  for (const batchCommitId of deltaBatchCommits.keys()) {
    for (const node of shapeQueries.listOutgoingNeighborsByEdgeType(
      fieldId,
      shapes.EdgeType.CreatedIn
    ).results) {
      if (node.result.id === batchCommitId) return { ...results, added: true };
    }
  }

  // This will not deal with array item changes
  for (const batchCommitId of deltaBatchCommits.keys()) {
    for (const node of shapeQueries.listOutgoingNeighborsByEdgeType(
      fieldId,
      shapes.EdgeType.UpdatedIn
    ).results) {
      if (node.result.id === batchCommitId)
        return { ...results, changed: true };
    }
  }

  // If a field is an array, there may be changes related to the shape but not
  // the field itself.
  return checkForArrayChanges(
    shapeQueries,
    deltaBatchCommits,
    results,
    shapeId
  );
}

export function getArrayChanges(
  shapeQueries: shapes.GraphQueries,
  shapeId: string,
  sinceBatchCommitId?: string
): ChangeResult {
  const results = {
    added: false,
    changed: false,
  };

  const deltaBatchCommits = getDeltaBatchCommits(
    shapeQueries,
    sinceBatchCommitId
  );

  return checkForArrayChanges(
    shapeQueries,
    deltaBatchCommits,
    results,
    shapeId
  );
}

function checkForArrayChanges(
  shapeQueries: shapes.GraphQueries,
  deltaBatchCommits: any,
  results: ChangeResult,
  shapeId: string
): ChangeResult {
  for (const batchCommitId of deltaBatchCommits.keys()) {
    for (const node of shapeQueries.listOutgoingNeighborsByEdgeType(
      shapeId,
      shapes.EdgeType.CreatedIn
    ).results) {
      if (node.result.id === batchCommitId) return { ...results, added: true };
    }
  }

  // This will not deal with array item changes
  for (const batchCommitId of deltaBatchCommits.keys()) {
    for (const node of shapeQueries.listOutgoingNeighborsByEdgeType(
      shapeId,
      shapes.EdgeType.UpdatedIn
    ).results) {
      if (node.result.id === batchCommitId)
        return { ...results, changed: true };
    }
  }

  return results;
}

type ChangeResult = {
  added: boolean;
  changed: boolean;
};

// TODO: use the endpointQueries one below
function getDeltaBatchCommits(
  shapeQueries: shapes.GraphQueries,
  sinceBatchCommitId?: string
): any {
  let sortedBatchCommits = shapeQueries
    .listNodesByType(shapes.NodeType.BatchCommit)
    .results.sort((a: any, b: any) => {
      return a.result.data.createdAt < b.result.data.createdAt ? 1 : -1;
    });
  const sinceBatchCommit: any = shapeQueries.findNodeById(sinceBatchCommitId!)!;
  const deltaBatchCommits = new Map();
  (sinceBatchCommitId
    ? sortedBatchCommits.filter(
        (batchCommit: any) =>
          batchCommit.result.data.createdAt >
          sinceBatchCommit!.result.data.createdAt
      )
    : sortedBatchCommits
  ).forEach((batchCommit: any) => {
    deltaBatchCommits.set(batchCommit.result.id, batchCommit);
  });
  return deltaBatchCommits;
}

export type ContributionsProjection = Record<string, Record<string, string>>;

export function getContributionsProjection(
  spec: any,
  opticEngine: any
): ContributionsProjection {
  return JSON.parse(opticEngine.get_contributions_projection(spec));
}
