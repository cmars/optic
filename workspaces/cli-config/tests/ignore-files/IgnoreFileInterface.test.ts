import { IgnoreFileHelper } from '../../src/helpers/ignore-file-interface';
import * as path from 'path';
import fs from 'fs-extra';

test('can parse rules from file, with Optic yml ones included', async () => {
  const ignoreHelper = new IgnoreFileHelper(
    path.join(__dirname, 'examples', 'ignore1'),
    path.join(__dirname, 'examples', 'optic1.yml')
  );

  const rules = await ignoreHelper.getCurrentIgnoreRules();
  console.log(rules);
  expect(rules).toMatchSnapshot();
});

test('can append new rule to file', async () => {
  const v2 = path.join(__dirname, 'examples', 'ignore2');
  await fs.copyFile(path.join(__dirname, 'examples', 'ignore1'), v2);

  const ignoreHelper = new IgnoreFileHelper(
    v2,
    path.join(__dirname, 'examples', 'optic1.yml')
  );

  await ignoreHelper.appendRule('POST /hello-world');
  const contents = (await fs.readFile(v2)).toString();
  await fs.unlink(v2);
  expect(contents).toMatchSnapshot();
});
