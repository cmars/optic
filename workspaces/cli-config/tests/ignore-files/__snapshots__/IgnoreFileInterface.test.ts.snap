// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can append new rule to file 1`] = `
"GET /images/(.*)
GET favicon.ico
GET bots.json
# These ignore methods, because methods...
OPTIONS *
HEAD *
POST /hello-world"
`;

exports[`can parse rules from file, with Optic yml ones included 1`] = `
Object {
  "allRules": Array [
    "GET /images/(.*)",
    "GET favicon.ico",
    "GET bots.json",
    "OPTIONS *",
    "HEAD *",
    "OPTIONS (.*)",
    "GET /main/resource",
  ],
  "ruleLines": Array [
    Object {
      "line": 0,
      "rule": "GET /images/(.*)",
    },
    Object {
      "line": 1,
      "rule": "GET favicon.ico",
    },
    Object {
      "line": 2,
      "rule": "GET bots.json",
    },
    Object {
      "line": 4,
      "rule": "OPTIONS *",
    },
    Object {
      "line": 5,
      "rule": "HEAD *",
    },
  ],
  "rulesFromYamlDeprecated": Array [
    "OPTIONS (.*)",
    "GET /main/resource",
  ],
}
`;
